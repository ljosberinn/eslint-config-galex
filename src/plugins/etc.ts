import type { RulesetCreator } from '../types';

export const createEtcPlugin: RulesetCreator = ({
  rules: customRules,
  ...dependencies
}) => ({
  ...createEtcRules(dependencies),
  ...customRules,
});

/**
 * @see https://github.com/cartant/eslint-plugin-etc
 *
 */
export const createEtcRules: RulesetCreator = ({
  typescript: { hasTypeScript },
}) => ({
  /**
   * forbids the assignment of returned, mutated arrays.
   *
   * @see https://github.com/cartant/eslint-plugin-etc/blob/main/docs/rules/no-assign-mutated-array.md
   */
  'etc/no-assign-mutated-array': 'error',

  /**
   * forbids commented-out code.
   *
   * @see https://github.com/cartant/eslint-plugin-etc/blob/main/docs/rules/no-commented-out-code.md
   */
  'etc/no-commented-out-code': 'warn',

  /**
   * forbids the use of deprecated APIs.
   *
   * @see https://github.com/cartant/eslint-plugin-etc/blob/main/docs/rules/no-deprecated.md
   */
  'etc/no-deprecated': hasTypeScript ? 'off' : 'warn',

  /**
   * forbids the use of internal APIs.
   *
   * @see https://github.com/cartant/eslint-plugin-etc/blob/main/docs/rules/no-internal.md
   */
  'etc/no-internal': 'warn',

  /**
   * Forbids type parameters without inference sites and type parameters that don't add type safety to declarations.
   *
   * @see https://github.com/cartant/eslint-plugin-etc/blob/main/docs/rules/no-misused-generics.md
   */
  'etc/no-misused-generics': 'warn',

  /**
   * forbids the assignment of returned, mutated arrays.
   *
   * @see https://github.com/cartant/eslint-plugin-etc/blob/main/docs/rules/no-assign-mutated-array.md
   */
  'etc/underscore-internal': 'warn',
});
