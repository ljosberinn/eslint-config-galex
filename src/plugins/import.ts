import type { RulesetCreator } from '../types';

export const createImportPlugin: RulesetCreator = ({
  rules: customRules = {},
  ...dependencies
}) => ({
  ...createImportRules(dependencies),
  ...customRules,
});

/**
 * @see https://github.com/benmosher/eslint-plugin-import
 *
 */
export const createImportRules: RulesetCreator = ({
  typescript: { hasTypeScript },
  react: { isCreateReactApp },
}) => ({
  'import/default': 'warn',
  /**
   * reports any dynamic imports without a webpackChunkName specificied
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/dynamic-import-chunkname.md
   */
  'import/dynamic-import-chunkname': 'warn',

  /**
   * warns about colliding exports
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/export.md
   */
  'import/export': hasTypeScript ? 'off' : 'error',

  /**
   * off because too opinonated
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/exports-last.md
   */
  'import/exports-last': 'off',

  /**
   * off because workspace sensitive
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/extensions.md
   */
  'import/extensions': 'off',

  /**
   * reports any imports that come after non-import statements
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/first.md
   */
  'import/first': 'error',

  /**
   * off because too opinionated
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/group-exports.md
   */
  'import/group-exports': 'off',

  /**
   * off because arbitrary
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/max-dependencies.md
   */
  'import/max-dependencies': 'off',

  /**
   * warns about nonexistent imports
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/named.md
   */
  'import/named': hasTypeScript ? 'off' : 'error',

  /**
   * warns about nonexistent properties on namespaces
   * similar to import/named
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/namespace.md
   * @see import/named
   */
  'import/namespace': hasTypeScript ? 'off' : 'error',

  /**
   * enforces a new line right after the last import of a file
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/newline-after-import.md
   */
  'import/newline-after-import': 'warn',

  /**
   * off because arbitrary & workspace sensitive
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-absolute-path.md
   */
  'import/no-absolute-path': 'off',

  /**
   * prevents usage of amd import/exports
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-amd.md
   */
  'import/no-amd': 'error',

  /**
   * prefer named exports. prefer named default exports
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-anonymous-default-export.md
   */
  'import/no-anonymous-default-export': 'error',

  /**
   * off because workspace sensitive
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-commonjs.md
   */
  'import/no-commonjs': 'off',

  /**
   * depects potential cyclical imports
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-cycle.md
   */
  'import/no-cycle': isCreateReactApp
    ? 'warn'
    : [
        'warn',
        {
          ignoreExternal: true,
          maxDepth: 5,
        },
      ],

  /**
   * any module should exclusively contain named exports
   * when unavoidable due to limitations, disable the warning for this line
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-default-export.md
   */
  'import/no-default-export': 'warn',

  /**
   * prevents the use of deprecated methods
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-deprecated.md
   */
  'import/no-deprecated': hasTypeScript ? 'off' : 'warn',

  /**
   * prevents importing the same path multiple times
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-duplicates.md
   * @see no-duplicate-imports
   */
  'import/no-duplicates': hasTypeScript ? 'off' : 'warn',

  /**
   * pervents use of dynamic `require`s. when necessary, disable the rule.
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-dynamic-require.md
   */
  'import/no-dynamic-require': 'warn',

  /**
   * prevents using dependencies not defined in package.json
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-extraneous-dependencies.md
   */
  'import/no-extraneous-dependencies': 'warn',

  /**
   * forbids mixing module.exports with import syntax
   *
   * off because randomly breaks
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-import-module-exports.md
   */
  'import/no-import-module-exports': 'off',

  /**
   * off because intended by some packages, e.g. `react-icons`, `next`, `msw`
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-internal-modules.md
   */
  'import/no-internal-modules': 'off',

  /**
   * prevents mutable exports
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-mutable-exports.md
   */
  'import/no-mutable-exports': 'error',

  /**
   * prevents importing named exports by the name of the default export
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-named-as-default.md
   */
  'import/no-named-as-default': 'off',

  /**
   * prevents use of an exported name as a property on the default export
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-named-as-default-member.md
   */
  'import/no-named-as-default-member': 'off',

  /**
   * prefer directly renaming defaults when importing
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-named-default.md
   */
  'import/no-named-default': 'error',

  /**
   * off because named exports should be the standard
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-named-export.md
   */
  'import/no-named-export': 'off',

  /**
   * prevents `import * as foo` syntax, use `import foo` instead
   *
   * off in tests!
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-namespace.md
   */
  'import/no-namespace': 'error',

  /**
   * off because workspace sensitive
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-nodejs-modules.md
   */
  'import/no-nodejs-modules': 'off',

  /**
   * off because workspace sensitive
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-relative-packages.md
   */
  'import/no-relative-packages': 'off',

  /**
   * prevents imports from parent folders
   *
   * off because not necessarily a bad pattern
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-relative-parent-imports.md
   */
  'import/no-relative-parent-imports': 'off',

  /**
   * off because workspace sensitive
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-restricted-paths.md
   */
  'import/no-restricted-paths': 'off',

  /**
   * prevent importing from the same file
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-self-import.md
   */
  'import/no-self-import': 'error',

  /**
   * hints unused imoprts. off because both VSCode and TS already hint this
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unassigned-import.md
   */
  'import/no-unassigned-import': 'off',

  /**
   * ensures an imported module actually exists
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unresolved.md
   */
  'import/no-unresolved': hasTypeScript ? 'off' : 'warn',

  /**
   * reports modules without exports or `importe`d/`require`d.
   * prevents dead code
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-unused-modules.md
   */
  'import/no-unused-modules': 'warn',

  /**
   * off because its the slowest rule in this config
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-useless-path-segments.md
   */
  'import/no-useless-path-segments': 'off',

  /**
   * prevents using loader syntax in imports. use webpack config instead.
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-webpack-loader-syntax.md
   */
  'import/no-webpack-loader-syntax': 'error',

  /**
   * - groups imports
   * - alphabetically sorts them
   * - enforces new lines between groups
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md
   */
  'import/order': [
    'warn',
    {
      alphabetize: {
        order: 'asc',
      },
      groups: [
        ['builtin', 'external', 'internal'],
        ['unknown', 'parent', 'sibling'],
        'index',
      ],
      'newlines-between': 'always',
    },
  ],

  /**
   * off because named exports are preferred
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/prefer-default-export.md
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-default-export.md
   */
  'import/prefer-default-export': 'off',

  /**
   * warns if a module is probably a script
   *
   * off because I frankly don't understand the description and it warns this
   * entire repo
   *
   * @see https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/unambiguous.md
   */
  'import/unambiguous': 'off',
});
