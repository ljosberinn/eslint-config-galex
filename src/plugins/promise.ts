import type { RulesCreator, RulesetCreator } from '../types';

export const createPromiseRules: RulesetCreator = ({
  rules: customRules = {},
  ...dependencies
}) => ({
  ...getPromiseRules(dependencies),
  ...customRules,
});

/**
 * @see https://github.com/xjamundx/eslint-plugin-promise
 */
export const getPromiseRules: RulesCreator = ({
  typescript: { hasTypeScript },
}) => ({
  /**
   * off because superseded by `promise/catch-or-return`
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/always-return.md
   * @see promise/catch-or-return
   */
  'promise/always-return': 'off',

  /**
   * off because too opinionated; not gonna pull in a dependency just for that
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/avoid-new.md
   */
  'promise/avoid-new': 'off',

  /**
   * enforces that all promises either be returned or handled
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/catch-or-return.md
   */
  'promise/catch-or-return': hasTypeScript
    ? 'off'
    : [
        'error',
        {
          allowFinally: true,
          allowThen: true,
        },
      ],

  /**
   * off because too opinionated; not gonna pull in a dependency just for that
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/no-callback-in-promise.md
   */
  'promise/no-callback-in-promise': 'off',

  /**
   * off because only appliesto ES5
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/no-native.md
   */

  'promise/no-native': 'off',
  /**
   * prefer async/await; use Promise.all where possible
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/no-nesting.md
   */
  'promise/no-nesting': 'warn',

  /**
   * use Promise.resolve/Promise.reject when possible
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/no-new-statics.md
   */
  'promise/no-new-statics': 'error',

  /**
   * enforces not using promises in callbacks
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/no-promise-in-callback.md
   */
  'promise/no-promise-in-callback': 'warn',

  /**
   * disallows returning in finally as its unhandled
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/no-return-in-finally.md
   */
  'promise/no-return-in-finally': 'warn',

  /**
   * avoid wrapping values in Promises when not needed
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/no-return-wrap.md
   */
  'promise/no-return-wrap': 'error',

  /**
   * disallows renaming callbacks within `new Promise`
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/param-names.md
   */
  'promise/param-names': 'warn',

  /**
   * try to promisify functions instead of passing callbacks around
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/prefer-await-to-callbacks.md
   */
  'promise/prefer-await-to-callbacks': 'warn',

  /**
   * prefer using async/await
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/prefer-await-to-then.md
   */
  'promise/prefer-await-to-then': 'warn',

  /**
   * ensures correct amount of required params passed to promises
   *
   * @see https://github.com/xjamundx/eslint-plugin-promise/blob/master/docs/rules/valid-params.md
   */
  'promise/valid-params': hasTypeScript ? 'off' : 'error',
});
