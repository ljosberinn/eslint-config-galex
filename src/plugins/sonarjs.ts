import type { RulesCreator, RulesetCreator } from '../types';

export const createSonarjsRules: RulesetCreator = ({
  rules: customRules = {},
  ...dependencies
}) => ({
  ...getSonarJsRules(dependencies),
  ...customRules,
});

/**
 * @see https://github.com/SonarSource/eslint-plugin-sonarjs
 */
export const getSonarJsRules: RulesCreator = ({
  typescript: { hasTypeScript },
}) => ({
  /**
   * prevents creeping complexity. consider alternative approach.
   *
   * @default 15
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/cognitive-complexity.md
   */
  'sonarjs/cognitive-complexity': 'warn',

  /**
   * requires exhaustive elseif-else
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/elseif-without-else.md
   */
  'sonarjs/elseif-without-else': 'warn',

  /**
   * prevents endless switches. consider alternative approach. limit upped to 15.
   *
   * @default 15
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/max-switch-cases.md
   */
  'sonarjs/max-switch-cases': ['error', 15],

  /**
   * prevents duplicate branches in control statements
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-all-duplicated-branches.md
   */
  'sonarjs/no-all-duplicated-branches': 'error',

  /**
   * merges nested ifs that have nothing in between
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-collapsible-if.md
   */
  'sonarjs/no-collapsible-if': 'error',

  /**
   * prevents impossible or implausible array length checks
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-collection-size-mischeck.md
   */
  'sonarjs/no-collection-size-mischeck': 'warn',

  /**
   * prevents use of strings that should be extracted as constants
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-duplicate-string.md
   */
  'sonarjs/no-duplicate-string': 'error',

  /**
   * prevents code duplication in control statements
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-duplicated-branches.md
   */
  'sonarjs/no-duplicated-branches': 'error',

  /**
   * prevents accidental unconditional overrides of a previously hardcoded value
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-element-overwrite.md
   */
  'sonarjs/no-element-overwrite': 'error',

  /**
   * prevents accessing empty collections
   *
   * off with TS as it handles this itself
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-empty-collection.md
   */
  'sonarjs/no-empty-collection': hasTypeScript ? 'off' : 'warn',

  /**
   * prevents unecessary arguments
   *
   * off when using typescript, already taken care of
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-extra-arguments.md
   */
  'sonarjs/no-extra-arguments': hasTypeScript ? 'off' : 'error',

  /**
   * Boolean expressions should not be gratuitous
   *
   * off because taken care of by `unicorn/no-lonely-if`
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-gratuitous-expressions.md
   */
  'sonarjs/no-gratuitous-expressions': 'off',

  /**
   * prevents duplicate branches
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-identical-conditions.md
   */
  'sonarjs/no-identical-conditions': 'error',

  /**
   * prevents unecessary & potentially unecessarily complex code
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-identical-expressions.md
   */
  'sonarjs/no-identical-expressions': 'error',

  /**
   * prevents duplicating function bodies
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-identical-functions.md
   */
  'sonarjs/no-identical-functions': 'error',

  /**
   * Return values from functions without side effects should not be ignored
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-ignored-return.md
   */
  'sonarjs/no-ignored-return': 'error',

  /**
   * prevents unecessary complexity
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-inverted-boolean-check.md
   */
  'sonarjs/no-inverted-boolean-check': 'error',

  /**
   * prefer not nesting switch
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-nested-switch.md
   */
  'sonarjs/no-nested-switch': 'warn',

  /**
   * prevents nesting templ literals
   *
   * off because opinitonated. not exactly a nice thing to do but... not terribly bad either
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-nested-template-literals.md
   */
  'sonarjs/no-nested-template-literals': 'off',

  /**
   * prevents unecessary loops
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-one-iteration-loop.md
   */
  'sonarjs/no-one-iteration-loop': 'error',

  /**
   * prevents use of `if(result === true)` in favor of `if(result)`
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-redundant-boolean.md
   */
  'sonarjs/no-redundant-boolean': 'error',

  /**
   * prevents unecessary `return`/`break`/`continue` statements
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-redundant-jump.md
   */
  'sonarjs/no-redundant-jump': 'error',

  /**
   * off because taken care of by prettier
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-same-line-conditional.md
   */
  'sonarjs/no-same-line-conditional': 'off',

  /**
   * prevents unecessary complexity
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-small-switch.md
   */
  'sonarjs/no-small-switch': 'warn',

  /**
   * prevents unecessary complexity
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-unused-collection.md
   */
  'sonarjs/no-unused-collection': 'error',

  /**
   * prevents bugs by expecting a return type
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-use-of-empty-return-value.md
   */
  'sonarjs/no-use-of-empty-return-value': 'error',

  /**
   * prevents unecessary complexity
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-useless-catch.md
   */
  'sonarjs/no-useless-catch': 'error',

  /**
   * prevents usage of possibly unintended "operators"
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/non-existent-operator.md
   */
  'sonarjs/non-existent-operator': 'warn',

  /**
   * prevents unecessary code
   *
   * warning because autofixable
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/prefer-immediate-return.md
   */
  'sonarjs/prefer-immediate-return': 'warn',

  /**
   * prevents bad practice
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/prefer-object-literal.md
   */
  'sonarjs/prefer-object-literal': 'error',

  /**
   * prevents unecessary complexity
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/prefer-single-boolean-return.md
   */
  'sonarjs/prefer-single-boolean-return': 'error',

  /**
   * prefers `while` over `for` for "infinite" loops
   *
   * @see https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/prefer-while.md
   */
  'sonarjs/prefer-while': 'error',
});
