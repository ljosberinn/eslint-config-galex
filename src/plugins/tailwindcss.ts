import type { RulesetCreator } from '../types';

export const createTailwindPlugin: RulesetCreator = ({
  rules: customRules,
  ...dependencies
}) => ({
  ...createTailwindRules(dependencies),
  ...customRules,
});

/**
 * @see https://github.com/francoismassart/eslint-plugin-tailwindcss
 *
 */
export const createTailwindRules: RulesetCreator = () => ({
  /**
   * order classnames for consistency and it makes merge conflict a bit easier to resolve
   *
   * @see https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/classnames-order.md
   */
  'tailwindcss/classnames-order': 'warn',

  /**
   * make sure to use negative arbitrary values classname without the negative classname e.g. -top-[5px] should become top-[-5px]
   *
   * @see https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/enforces-negative-arbitrary-values.md
   */
  'tailwindcss/enforces-negative-arbitrary-values': 'warn',

  /**
   * merge multiple classnames into shorthand if possible e.g. mx-5 my-5 should become m-5
   *
   * @see https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/enforces-shorthand.md
   */
  'tailwindcss/enforces-shorthand': 'warn',

  /**
   * for easy upgrade from Tailwind CSS v2 to v3. Warning: at the moment you should temporary turn off the no-custom-classname rule if you want to see the warning from migration-from-tailwind-2
   *
   * @see https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/migration-from-tailwind-2.md
   */
  'tailwindcss/migration-from-tailwind-2': 'warn',

  /**
   * forbid using arbitrary values in classnames (turned off by default)
   *
   * @see https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/no-arbitrary-value.md
   */
  'tailwindcss/no-arbitrary-value': 'off',

  /**
   * only allow classnames from Tailwind CSS and the values from the whitelist option
   *
   * @see https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/no-custom-classname.md
   */
  'tailwindcss/no-custom-classname': 'warn',

  /**
   * e.g. avoid p-2 p-3, different Tailwind CSS classnames (pt-2 & pt-3) but targeting the same property several times for the same variant.
   *
   * @see https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/no-contradicting-classname.md
   */
  'tailwindcss/no-contradicting-classname': 'error',
});
