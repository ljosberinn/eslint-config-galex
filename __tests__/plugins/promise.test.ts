import packageJson from '../../package.json';
import { createPromisePlugin } from '../../src/plugins/promise';
import tsConfig from '../../tsconfig.json';
import { defaultProject } from '../shared';

test('allows passing rules', () => {
  const ruleName = 'foo';
  const ruleValue = 'off';

  const result = createPromisePlugin({
    ...defaultProject,
    rules: {
      [ruleName]: ruleValue,
    },
  });

  expect(result[ruleName]).toBe(ruleValue);
});

test('with typescript', () => {
  expect(
    createPromisePlugin({
      ...defaultProject,
      typescript: {
        hasTypeScript: true,
        version: packageJson.dependencies.typescript,
        // @ts-expect-error false positive
        config: tsConfig,
      },
    })
  ).toMatchSnapshot();
});
