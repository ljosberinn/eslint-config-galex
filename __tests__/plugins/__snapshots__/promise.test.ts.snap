// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with typescript 1`] = `
{
  "promise/always-return": "off",
  "promise/avoid-new": "off",
  "promise/catch-or-return": "off",
  "promise/no-callback-in-promise": "off",
  "promise/no-native": "off",
  "promise/no-nesting": "warn",
  "promise/no-new-statics": "error",
  "promise/no-promise-in-callback": "warn",
  "promise/no-return-in-finally": "warn",
  "promise/no-return-wrap": "error",
  "promise/param-names": "warn",
  "promise/prefer-await-to-callbacks": "warn",
  "promise/prefer-await-to-then": "warn",
  "promise/valid-params": "off",
}
`;
